From d5a5241ef7d9af63f2f3423105cbf0eaae6c6440 Mon Sep 17 00:00:00 2001
From: Fancy Fang <chen.fang@nxp.com>
Date: Fri, 10 Aug 2018 12:22:18 +0800
Subject: [PATCH 4380/5242] MLK-19199 drm/imx: lcdif: replace FB width usage
 for cropping

commit  3622c4b5d81bafed4026f6773fe95a1baa2ef82b from
https://source.codeaurora.org/external/imx/linux-imx.git

According to the comments of 'struct drm_framebuffer', its
'width' field refers to the logical width of the visible
area of the framebuffer. This may be unequal to the total
pixels number of a line. So use the 'pitches' field to
replace 'width' for the horizontal cropping feature.

Signed-off-by: Fancy Fang <chen.fang@nxp.com>
(cherry picked from commit 9a2bbbf971ed79b32ae1c7da2d62b8a72f3ccffd)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/imx/lcdif/lcdif-plane.c |   13 ++++++++-----
 1 file changed, 8 insertions(+), 5 deletions(-)

diff --git a/drivers/gpu/drm/imx/lcdif/lcdif-plane.c b/drivers/gpu/drm/imx/lcdif/lcdif-plane.c
index 91bd6fc..6662ba5 100644
--- a/drivers/gpu/drm/imx/lcdif/lcdif-plane.c
+++ b/drivers/gpu/drm/imx/lcdif/lcdif-plane.c
@@ -94,12 +94,12 @@ static int lcdif_plane_atomic_check(struct drm_plane *plane,
 	if (!plane_state->visible)
 		return -EINVAL;
 
-	/* force 'mode_changed' when fb width changed, since
+	/* force 'mode_changed' when fb pitches changed, since
 	 * the pitch related registers configuration of LCDIF
 	 * can not be done when LCDIF is running.
 	 */
 	if (old_fb && likely(!crtc_state->mode_changed)) {
-		if (old_fb->width != fb->width)
+		if (old_fb->pitches[0] != fb->pitches[0])
 			crtc_state->mode_changed = true;
 	}
 
@@ -114,7 +114,7 @@ static void lcdif_plane_atomic_update(struct drm_plane *plane,
 	struct drm_plane_state *state = plane->state;
 	struct drm_framebuffer *fb = state->fb;
 	struct drm_gem_cma_object *gem_obj = NULL;
-	u32 fb_addr, src_off, src_w, fb_idx;
+	u32 fb_addr, src_off, src_w, fb_idx, cpp, stride;
 	bool crop;
 
 	/* plane and crtc is disabling */
@@ -146,11 +146,14 @@ static void lcdif_plane_atomic_update(struct drm_plane *plane,
 
 	/* config horizontal cropping if crtc needs modeset */
 	if (unlikely(drm_atomic_crtc_needs_modeset(state->crtc->state))) {
+		cpp = fb->format->cpp[0];
+		stride = DIV_ROUND_UP(fb->pitches[0], cpp);
+
 		src_w = state->src_w >> 16;
 		WARN_ON(src_w > fb->width);
 
-		crop  = src_w != fb->width ? true : false;
-		lcdif_set_fb_hcrop(lcdif, src_w, fb->width, crop);
+		crop  = src_w != stride ? true : false;
+		lcdif_set_fb_hcrop(lcdif, src_w, stride, crop);
 	}
 
 	lcdif_enable_controller(lcdif);
-- 
1.7.9.5

